### PR TITLE
feat: add keyless backup into onboarding flow

### PR DESCRIPTION
**Pull request by MuckT**
### Description

Adds the keyless backup flow into onboarding with a redesigned `SignInWithEmail.tsx` screen specifically for onboarding. Note the onboarding background colors are fixed in #5592 and the below videos use a throwaway number and seed phrase.

| iOS CAB | iOS Recovery Phrase | iOS Skip |
| ----- | ----- | ----- |
| <video src="https://github.com/valora-inc/wallet/assets/26950305/d7f2a3bb-315f-4962-8db1-114e2c686d6b" /> | <video src="https://github.com/valora-inc/wallet/assets/26950305/86beca60-4d9f-49f8-ac2c-ebd3c84712b7" /> | <video src="https://github.com/valora-inc/wallet/assets/26950305/1672d93a-008d-47e8-b149-13943f98e6fa" /> |

| Android CAB | Android Recovery Phrase | Android Skip |
| ------ | ----- | ----- |
| <video src="https://github.com/valora-inc/wallet/assets/26950305/3ef7826d-ec10-4a76-8605-e4b1413bf9ff" /> | <video src="https://github.com/valora-inc/wallet/assets/26950305/37e929c8-f0b3-487d-9a4d-f63e18f412ab" /> | <video src="https://github.com/valora-inc/wallet/assets/26950305/d5d494a8-a343-45cb-9a5e-29b3b2b69e5c" /> | 

### Test plan

- [x] Tested locally on iOS
- [x] Tested locally on Android
- [x] Unit tests updated

### Related issues

- Fixes ACT-1219

### Backwards compatibility

Yes

### Network scalability

N/A

### CODE REVIEW

File: src/navigator/types.tsx [hunk 1]
```diff
@@ -56,7 +56,8 @@ export type StackParamList = {
   [Screens.AccountKeyEducation]:
     | undefined
     | {
-        nextScreen: keyof StackParamList
+        nextScreen?: keyof StackParamList
+        origin?: 'cabOnboarding' | 'onboarding' | 'settings'
```
Comment (MuckT) [inline src/navigator/types.tsx:60]: I think this can be derived instead 🤔

File: src/onboarding/steps.ts
Comment (MuckT) [inline src/onboarding/steps.ts:1]: Updating the steps tests proves a bit difficult as I am using two screens in a single step and a different screens step count in `OnboardingRecoveryPhrase.tsx`; I am open to alternatives.

Comment (MuckT) [reply src/onboarding/steps.ts:1]: It looks like we do some references to steps of a different screen already in [`OnboardingRecoveryPhrase.tsx`](https://github.com/valora-inc/wallet/blob/main/src/onboarding/registration/OnboardingRecoveryPhrase.tsx#L37) so this may be a non-issue.

File: src/navigator/types.tsx (hunk 1 above)
Comment (MuckT) [reply src/navigator/types.tsx]: I ended up using `origin?: 'cabOnboarding'` so this could probably be a boolean; however, it might be more extensible as is.

File: src/onboarding/steps.ts [hunk 2]
```diff
@@ -243,6 +264,34 @@ export function _getStepInfo({ firstScreenInStep, navigator, dispatch, props }:
...
+          if (showRecoveryPhraseEducation) {
+            navigate(Screens.AccountKeyEducation, {
+              origin: 'cabOnboarding',
+            })
+          } else if (skipVerification || numberAlreadyVerifiedCentrally) {
+            dispatch(setHasSeenVerificationNux(true))
+            finishOnboarding(Screens.ChooseYourAdventure)
+          } else {
+            // DO NOT CLEAR NAVIGATION STACK HERE - breaks restore flow on initial app open in native-stack v6
+            navigate(Screens.VerificationStartScreen)
```
Comment (satish-ravi) [inline src/onboarding/steps.ts:279]: In the videos, looks like after backup is successful, the flow navigates to home directly, is that expected? Per this, looks like it should go to phone verification / CYA?

File: src/keylessBackup/SignInWithEmail.tsx [hunk 3]
```diff
@@ -1,27 +1,78 @@
...
+  bottomSheetRef: React.RefObject<BottomSheetRefType>
+}) {
+  const { t } = useTranslation()
+
+  return (
+    <BottomSheet
+      forwardedRef={bottomSheetRef}
+      title={t('signInWithEmail.bottomSheet.title')}
+      titleStyle={styles.bottomSheetTitle}
+      testId="KeylessBackupSignInWithEmail/HelpInfoBottomSheet"
```
Comment (satish-ravi) [inline src/keylessBackup/SignInWithEmail.tsx:51]: nit: this is not help info right? could use a better name.

File: src/keylessBackup/SignInWithEmail.tsx [hunk 4]
```diff
@@ -124,8 +245,15 @@ const styles = StyleSheet.create({
   scrollContainer: {
     padding: Spacing.Thick24,
   },
+  header: {
+    paddingHorizontal: variables.contentPadding,
```
Comment (satish-ravi) [inline src/keylessBackup/SignInWithEmail.tsx:249]: this is 16 right? seems like the scroll container uses 24, should this be consistent?

File: src/onboarding/registration/OnboardingRecoveryPhrase.tsx [hunk 5]
```diff
@@ -32,9 +32,13 @@ import Logger from 'src/utils/Logger'
 
 type Props = NativeStackScreenProps<StackParamList, Screens.OnboardingRecoveryPhrase>
 
-function OnboardingRecoveryPhrase({ navigation }: Props) {
+function OnboardingRecoveryPhrase({ navigation, route }: Props) {
   const onboardingProps = useSelector(onboardingPropsSelector)
-  const { step, totalSteps } = getOnboardingStepValues(Screens.ProtectWallet, onboardingProps)
+  // Use a lower step count for CAB onboarding
```
[the remark this answers is not in the document]
Comment (MuckT) [reply src/onboarding/registration/OnboardingRecoveryPhrase.tsx:37]: It's similar to what was done in [here](https://github.com/valora-inc/wallet/blob/main/src/onboarding/registration/OnboardingRecoveryPhrase.tsx#L37 ) where we use a different screen to have a consistent step count.

File: src/onboarding/steps.ts (hunk 2 above)
Comment (MuckT) [reply src/onboarding/steps.ts:279]: This is fixed in #5611.

File: src/keylessBackup/SignInWithEmail.tsx (hunk 3 above)
Comment (MuckT) [reply src/keylessBackup/SignInWithEmail.tsx]: Fixed in 0320f30820f7ec8c5b4d92079b92b4780f717ac4!

File: src/keylessBackup/SignInWithEmail.tsx (hunk 4 above)
Comment (MuckT) [reply src/keylessBackup/SignInWithEmail.tsx:249]: I tried to be consistent with headers in the previous screens which appear to be 16.

File: src/keylessBackup/types.ts [hunk 6]
```diff
@@ -21,5 +21,6 @@ export enum KeylessBackupDeleteStatus {
 
 export enum KeylessBackupOrigin {
   Onboarding = 'Onboarding',
+  Restore = 'Restore',
```
[the remark this answers is not in the document]
Comment (MuckT) [reply src/keylessBackup/types.ts]: Fixed in acb95aeb2b1bcadbe28e6b7ee2d0975595437bee!

File: src/keylessBackup/SignInWithEmail.tsx [hunk 7]
```diff
@@ -30,6 +81,43 @@ function SignInWithEmail({ route }: Props) {
   const { authorize, getCredentials, clearCredentials } = useAuth0()
   const { keylessBackupFlow, origin } = route.params
   const [loading, setLoading] = useState(false)
+  const onboardingProps = useSelector(onboardingPropsSelector)
+  const { step, totalSteps } = getOnboardingStepValues(Screens.SignInWithEmail, onboardingProps)
+  const { bottom } = useSafeAreaInsets()
+  const insetsStyle = {
+    paddingBottom: Math.max(0, 40 - bottom),
+  }
+  const isOnboarding = origin === KeylessBackupOrigin.Onboarding
```
[the remark this answers is not in the document]
Comment (MuckT) [reply src/keylessBackup/SignInWithEmail.tsx]: Fixed in 0224ecb1f1e9cd55dd048d4183db93ff806b0e0e!

File: src/components/header/CustomHeader.tsx [hunk 8]
```diff
@@ -1,17 +1,23 @@
...
+  subTitle?: string | React.ReactNode
   style?: StyleProp<ViewStyle>
 }
 
-function CustomHeader({ left, right, title, style }: Props) {
-  const titleComponent =
-    typeof title === 'string' ? <Text style={headerStyles.headerTitle}>{title}</Text> : title
+function CustomHeader({ left, right, title, style, subTitle }: Props) {
+  const titleComponent = subTitle ? (
+    <HeaderTitleWithSubtitle title={title} subTitle={subTitle} />
```
[the remark this answers is not in the document]
Comment (MuckT) [reply src/components/header/CustomHeader.tsx]: Fixed in 32b452ab8cb7feb0b2156714719bb71f6bade8ad!

File: src/keylessBackup/SignInWithEmail.tsx [hunk 9]
```diff
@@ -30,6 +81,43 @@ function SignInWithEmail({ route }: Props) {
...
+  const onPressSkip = () => {
+    ValoraAnalytics.track(KeylessBackupEvents.cab_sign_in_with_email_screen_skip, {
+      keylessBackupFlow,
+      origin,
+    })
+    goToNextOnboardingScreen({
+      firstScreenInCurrentStep: Screens.SignInWithEmail,
+      onboardingProps,
+    })
+  }
```
[the remark this answers is not in the document]
Comment (MuckT) [reply src/keylessBackup/SignInWithEmail.tsx]: Changed in 685f385d9262082a6bc30e98b598850b63d3ffbc!

File: src/components/header/CustomHeader.tsx [hunk 10]
```diff
@@ -6,12 +6,14 @@ interface Props {
   left?: React.ReactNode
   right?: React.ReactNode
   title?: React.ReactNode | string
+  subTitle?: string | React.ReactNode
```
Comment (satish-ravi) [inline src/components/header/CustomHeader.tsx:9]: remove?

File: src/onboarding/steps.ts [hunk 11]
```diff
@@ -243,6 +264,34 @@ export function _getStepInfo({ firstScreenInStep, navigator, dispatch, props }:
...
+          }
+        },
+      }
+    case Screens.AccountKeyEducation:
+      return {
+        next: () => {
+          navigate(Screens.OnboardingRecoveryPhrase)
+        },
+      }
+    case Screens.OnboardingRecoveryPhrase:
```
Comment (satish-ravi) [inline src/onboarding/steps.ts:289]: are these considered individual steps in onboarding? looking at the code in `OnboardingRecoveryPhrase`, it seems like it uses `SignInWithEmail` as the "firstScreenInCurrentStep", so I am not sure if we need these here. Wonder if we could just navigate directly from the SignInWithEmail screen, instead of considering them as a separate `step`

Comment (MuckT) [reply src/onboarding/steps.ts]: Good catch we don't need these listed since we use navigation to these screens directly; fixed in a2a2d07f2712d3e3f69cfd0ee84a5fc7bdee994d!

File: src/components/header/CustomHeader.tsx (hunk 10 above)
Comment (MuckT) [reply src/components/header/CustomHeader.tsx]: Removed in b93f45c73831ae54a9235201fcb627d7b502d89c!

File: src/onboarding/steps.ts [hunk 12]
```diff
@@ -243,6 +264,22 @@ export function _getStepInfo({ firstScreenInStep, navigator, dispatch, props }:
           }
         },
       }
+    case Screens.SignInWithEmail:
+      return {
+        next: () => {
+          if (showRecoveryPhraseEducation) {
+            navigate(Screens.AccountKeyEducation, {
+              origin: 'cabOnboarding',
+            })
```
Comment (satish-ravi) [inline src/onboarding/steps.ts:273]: I think we should remove this case and directly navigate to this screen from SignInWithEmail. Otherwise there is an edge case where our step counter fn could crash when it walks thru graph since `AccountKeyEducation` does not have a next step. Although practically this may never happen since `showRecoveryPhraseEducation` will be `false`, but anyways good to keep this clean.

File: src/keylessBackup/SignInWithEmail.tsx [hunk 13]
```diff
@@ -1,27 +1,99 @@
...
+}) {
+  const { t } = useTranslation()
+  const onboardingProps = useSelector(onboardingPropsSelector)
+  const onPressContinue = () => {
+    ValoraAnalytics.track(KeylessBackupEvents.cab_setup_recovery_phrase)
+    bottomSheetRef.current?.close()
+    goToNextOnboardingScreen({
+      firstScreenInCurrentStep: Screens.SignInWithEmail,
+      onboardingProps: { ...onboardingProps, showRecoveryPhraseEducation: true },
+    })
```
Comment (satish-ravi) [inline src/keylessBackup/SignInWithEmail.tsx:53]: per comment in steps, this can just navigate directly

File: src/components/header/CustomHeader.tsx [hunk 14]
```diff
@@ -12,6 +12,7 @@ interface Props {
 function CustomHeader({ left, right, title, style }: Props) {
   const titleComponent =
     typeof title === 'string' ? <Text style={headerStyles.headerTitle}>{title}</Text> : title
+
```
Comment (satish-ravi) [inline src/components/header/CustomHeader.tsx:15]: nit: unnecessary change

File: src/keylessBackup/SignInWithEmail.tsx (hunk 13 above)
Comment (MuckT) [reply src/keylessBackup/SignInWithEmail.tsx]: Fixed in 03854844f6b6f9834fda241e3761180f23644925!

**Review by satish-ravi — APPROVED** [review]
